### ISSUE TITLE
Out of Memory Error: MemoryError: Unable to allocate array with shape (249255, 249255) and data type float64

### ISSUE THREAD

**Issue by Austin-s-h** [lanagarmire/deepimpute#12]
Hi! I'm trying to run DeepImpute on scATAC-Seq data.
I've filtered my dataset to 'high-quality' cells with at least 5500 reads.
![image](https://user-images.githubusercontent.com/44502805/69276592-c70cb280-0bac-11ea-9724-6b13b29d05d9.png)
I've filtered my features (peaks) for those observed in >10 cells, leaving me with close to 250k. When I try to run impute on this, it crashes. 
```
Input dataset is 7706 cells (rows) and 249255 genes (columns)
First 3 rows and columns:
                    W_14793_15289  W_37170_37548  W_46846_47099
AAACGAAAGTAATGTG-3              0              0              0
AAACGAACAGATGGCA-2              0              0              0
AAACGAACATTGTGAC-4              0              0              0
23040 genes selected for imputation
```
```
---------------------------------------------------------------------------
MemoryError                               Traceback (most recent call last)
<ipython-input-6-27275eb9ee2f> in <module>
      7 # Crashed, let's try with 50% of the data to fit the network.
      8 
----> 9 multinet.fit(MACS_data,cell_subset=1,minVMR=0.5)

~/.local/lib/python3.7/site-packages/deepimpute/multinet.py in fit(self, raw, cell_subset, NN_lim, genes_to_impute, ntop, minVMR, mode)
    192                 genes_to_impute = np.concatenate((genes_to_impute, fill_genes))
    193 
--> 194         covariance_matrix = get_distance_matrix(raw)
    195 
    196         self.setTargets(raw.reindex(columns=genes_to_impute), mode=mode)

~/.local/lib/python3.7/site-packages/deepimpute/multinet.py in get_distance_matrix(raw)
     22     potential_pred = raw.columns[raw.std() > 0]
     23 
---> 24     covariance_matrix = pd.DataFrame(np.abs(np.corrcoef(raw.T.loc[potential_pred])),
     25                                      index=potential_pred,
     26                                      columns=potential_pred).fillna(0)

<__array_function__ internals> in corrcoef(*args, **kwargs)

~/miniconda3/lib/python3.7/site-packages/numpy/lib/function_base.py in corrcoef(x, y, rowvar, bias, ddof)
   2524         warnings.warn('bias and ddof have no effect and are deprecated',
   2525                       DeprecationWarning, stacklevel=3)
-> 2526     c = cov(x, y, rowvar)
   2527     try:
   2528         d = diag(c)

<__array_function__ internals> in cov(*args, **kwargs)

~/miniconda3/lib/python3.7/site-packages/numpy/lib/function_base.py in cov(m, y, rowvar, bias, ddof, fweights, aweights)
   2452     else:
   2453         X_T = (X*w).T
-> 2454     c = dot(X, X_T.conj())
   2455     c *= np.true_divide(1, fact)
   2456     return c.squeeze()

<__array_function__ internals> in dot(*args, **kwargs)

MemoryError: Unable to allocate array with shape (249255, 249255) and data type float64
```
Could you explain why the program is trying to operate on a matrix of all_genes x all_genes? I am running this on a server with ~200GB of memory. I can override the error with `echo 1 > /proc/sys/vm/overcommit_memory` but it actually uses it all and crashes. Any thoughts would be appreciated! If I'm understanding correctly, I cannot run the model on the identified subset and then apply that similarly to the rest of the genes afterwards, correct? Additionally, I realize that this is meant for scRNA-Seq but I figured it should be able to be applied to scATAC-Seq. 

On another note, is it possible to specify which normalization to use? Say, Square-root transform?

**Comment by Austin-s-h** [comment]
Filtering down to 125k features allowed this to run using ~140GB of RAM.

**Comment by Puumanamana** [comment]
Hi Austin,
Thanks for your detailed explanation. Unfortunately, the way it is now, deepimpute cannot handle that many genes (or it needs a high amount of memory for it). The reason is that it builds all the sub-networks in parallel (and not sequentially), which is fine with less than 30k genes. 
My advice would be to increase the --minVMR options from 0.5 to maybe 2, to select less genes for imputation. 
If you really want to impute everything, you can also do it manually using the API directly. For example something like:
```python
from math import ceil
import pandas as pd
from deepimpute.multinet import MultiNet

# data = ... your data
all_genes = data.columns.tolist()
subset_size = 10000
imputed = []

for i in range(ceil(len(all_genes)/subset_size)):
    gene_subset = all_genes[i*subset_size:(i+1)*subset_size]

    model = MultiNet()
    model.fit(data, genes_to_impute=gene_subset)
    imputed.append(model.predict(data))

imputed = pd.concat(imputed, axis=1)
```

Regarding the normalization, non is needed. DeepImpute takes care of it for you, and undo the normalization after.

**Comment by Austin-s-h** [comment]
Thank you for this! I'm currently testing a 300GB swap allocation on a RAID 0 array of NVMe drives, which may be able to handle the memory requirements to run everything in parallel. 

Does this code also account for selection of genes to impute? And what do you exactly mean by genes to impute? Does this mean that the model is trained on this subset as a representation of significantly variable features?

Furthermore, I do not believe the error was arising during the imputation of that selected amount of genes. For my example, it selected 23040 genes but was trying to build a matrix all genes x all genes, could you explain why this step is necessary when we only want to impute 23040 genes?

**Comment by Puumanamana** [comment]
- Actually, you're right I didn't look at the error closely enough. One of the preliminary step of deepImpute is to compute a gene x gene correlation matrix. This matrix is used to choose the inputs: for each output gene, we choose the 5 best correlated genes and use them as inputs. 
The issue here is that since there are 300k genes, the correlation matrix is 300k x 300k. I have added a parameter in deepImpute to select the top `n_pred` genes based on the Variance over Mean Ratio (VMR) criteria. So if you set n_pred to 50k, the matrix will be 50k x 50k. It's not optimal since you're leaving potential good predictors aside, but since the genes with the best VMR are selected, you should have the main ones.

- The `genes_to_impute` parameter takes a custom set of genes the user wants to impute, and will only impute those (but the predictors can be among the non-selected genes).

- Regarding the 23,040 genes that are chosen for imputation, is the default behaviour of the algorithm. It focuses on genes with a VMR above 0.5 (default) to focus on variable genes to impute. In other words, we are not targeting genes that have a low variability in relation to their average value. You can change this threshold with the `minVMR` parameter.

**Comment by govindbisen** [comment]
Is It solved? I am facing this problem in much lower level. i am just trying to create matrix
cosine_sim = linear_kernel(tfmatrix, tfmatrix) and getting this - MemoryError: Unable to allocate 15.4 GiB for an array with shape (45466, 45466) and data type float64

**Comment by danielAdama** [comment]
@govindbisen try installing a 64-bit python

**Comment by sharvari21** [comment]
I am also facing such an error while doing the ARIMA model. 
"MemoryError: Unable to allocate 25.5 GiB for an array with shape (58548, 58548) and data type float64
"
This is my error please help me to solve this!

**Comment by bcikili** [comment]
> I am also facing such an error while doing the ARIMA model.
> "MemoryError: Unable to allocate 25.5 GiB for an array with shape (58548, 58548) and data type float64
> "
> This is my error please help me to solve this!
after you make sure you have python 64bit version, you can try converting datatypes to smaller sized versions. float63 to float 32, or if possible even to smaller ones like np.uint8 

Pred['train'] = Pred['train'].astype(np.uint8,errors='ignore')

other than these there are not many fixing methods around. Unless you have hardware solutions

**Comment by danielAdama** [comment]
I agree with Burhan, if that doesn’t work then you will have to get a
laptop with higher performance in hardware

**Comment by e253503** [comment]
I also face with similar error: `MemoryError: Unable to allocate 2.02 GiB for an array with shape (542124951,) and data type float32`

**Comment by sistaseetaram** [comment]
has anyone been able to resolve this issue, please share it...

**Comment by lanagarmire** [comment]
Hi, this is not an issue of DeepImpute but an issue of the input file being
too big.

We are trying to write wrapper to allow large dataset upload.

Best
Lana

**Comment by dharaneeswar21** [comment]
i am also getting the same error 
while doing SARIMA model

MemoryError: Unable to allocate 1.09 GiB for an array with shape (27, 27, 201600) and data type float64

**Comment by gitanupam** [comment]
I had a similar issue (not with deepimpute but another Python library but same error with an array of 13 million elements) and increasing the RAM size on GCP solved it

**Comment by hanzigs** [comment]
I am having similar issue
```
MemoryError : Unable to allocate 6.51 MiB for an array with shape (40616, 21) and data type int64 (  File \"sklearn\\neighbors\\_binary_tree.pxi\", line 579, in sklearn.neighbors._kd_tree.NeighborsHeap.__init__\n  File \"sklearn\\neighbors\\_binary_tree.pxi\", line 1349, in sklearn.neighbors._kd_tree.BinaryTree.query\n  File \"C:\\Python\\Lib\\site-packages\\sklearn\\neighbors\\_base.py\", line 492, in _tree_query_parallel_helper\n    return tree.query(*args, **kwargs)\n  File \"C:\\Python\\Lib\\site-packages\\joblib\\parallel.py\", line 262, in <listcomp>\n    return [func(*args, **kwargs)\n  File \"C:\\Python\\Lib\\site-packages\\joblib\\parallel.py\", line 262, in __call__\n    return [func(*args, **kwargs)\n  File \"C:\\Python\\Lib\\site-packages\\joblib\\_parallel_backends.py\", line 595, in __call__\n    return self.func(*args, **kwargs)\n  File \"C:\\Python\\Lib\\multiprocessing\\pool.py\", line 125, in worker\n    result = (True, func(*args, **kwds))\n  File \"C:\\Python\\Lib\\multiprocessing\\pool.py\", line 771, in get\n    raise self._value\n  File \"C:\\Python\\Lib\\site-packages\\joblib\\parallel.py\", line 933, in retrieve\n    self._output.extend(job.get(timeout=self.timeout))\n  File \"C:\\Python\\Lib\\site-packages\\joblib\\parallel.py\", line 1054, in __call__\n    self.retrieve()\n  File \"C:\\Python\\Lib\\site-packages\\sklearn\\neighbors\\_base.py\", line 662, in kneighbors\n    chunked_results = Parallel(n_jobs, **parallel_kwargs)(\n  File \"C:\\Python\\Lib\\site-packages\\sklearn\\neighbors\\_lof.py\", line 271, in fit\n    self._distances_fit_X_, _neighbors_indices_fit_X_ = self.kneighbors(\n  File \"C:\\Python\\Lib\\site-packages\\pyod\\models\\lof.py\", line 173, in fit\n    self.detector_.fit(X=X, y=y)\n  File \"C:\\Python\\Lib\\site-packages\\pyod\\models\\feature_bagging.py\", line 284, in fit\n    estimator.fit(X[:, features])\n  File \"C:\\src\\scripts\\mlu.py\", line 721, in outC\n    fb_model = FeatureBagging(n_estimators=1000,\n  File \"C:\\src\\scripts\\mlt.py\", line 468, in train\n    df, no = outC(raw_df, target, param.outlier_percentage)\n===\n   at Python.Runtime.PyObject.Invoke(PyTuple args, PyDict kw)\r\n   at Python.Runtime.PyObject.InvokeMethod(String name, PyTuple args, PyDict kw)\r\n   at Python.Runtime.PyObject.TryInvokeMember(InvokeMemberBinder binder, Object[] args, Object& result)\r\n   at CallSite.Target(Closure , CallSite , Object , String , Object )\r\n   at System.Dynamic.UpdateDelegates.UpdateAndExecute3[T0,T1,T2,TRet](CallSite site, T0 arg0, T1 arg1, T2 arg2)\r\n   at Intuition.ML.Trainer.Run(Guid tenantId, Trainparameter) in C:\\src\\src\\Intuition.AutoML\\Implementation\\Trainer.cs:line 91)
```

**Comment by hanzigs** [comment]
What are the Checks needed, 
Python and Azure VM OS to be 64bit
and the only check is the RAM size, is it?
![image](https://user-images.githubusercontent.com/30790120/160506872-62a36f66-97cc-4a22-83fd-2675584587e6.png)
as per the Error, RAM went shot of 6.51MB
```
MemoryError : Unable to allocate 6.51 MiB for an array with shape (40616, 21) and data type int64 (  File \"sklearn\\neighbors\\_binary_tree.pxi\", line 579, in sklearn.neighbors._kd_tree.NeighborsHeap.__init__\n  File \"sklearn\\neighbors\\_binary_tree.pxi\", line 1349, in sklearn.neighbors._kd_tree.BinaryTree.query\n  File \"C:\\Python\\Lib\\site-packages\\sklearn\\neighbors\\_base.py\", line 492, in _tree_query_parallel_helper\n    return tree.query(*args, **kwargs)\n  File \"C:\\Python\\Lib\\site-packages\\joblib\\parallel.py\", line 262, in <listcomp>\n    return [func(*args, **kwargs)\n  File \"C:\\Python\\Lib\\site-packages\\joblib\\parallel.py\", line 262, in __call__\n    return [func(*args, **kwargs)\n  File \"C:\\Python\\Lib\\site-packages\\joblib\\_parallel_backends.py\", line 595, in __call__\n    return self.func(*args, **kwargs)\n  File \"C:\\Python\\Lib\\multiprocessing\\pool.py\", line 125, in worker\n    result = (True, func(*args, **kwds))\n  File \"C:\\Python\\Lib\\multiprocessing\\pool.py\", line 771, in get\n    raise self._value\n  File \"C:\\Python\\Lib\\site-packages\\joblib\\parallel.py\", line 933, in retrieve\n    self._output.extend(job.get(timeout=self.timeout))\n  File \"C:\\Python\\Lib\\site-packages\\joblib\\parallel.py\", line 1054, in __call__\n    self.retrieve()\n  File \"C:\\Python\\Lib\\site-packages\\sklearn\\neighbors\\_base.py\", line 662, in kneighbors\n    chunked_results = Parallel(n_jobs, **parallel_kwargs)(\n  File \"C:\\Python\\Lib\\site-packages\\sklearn\\neighbors\\_lof.py\", line 271, in fit\n    self._distances_fit_X_, _neighbors_indices_fit_X_ = self.kneighbors(\n  File \"C:\\Python\\Lib\\site-packages\\pyod\\models\\lof.py\", line 173, in fit\n    self.detector_.fit(X=X, y=y)\n  File \"C:\\Python\\Lib\\site-packages\\pyod\\models\\feature_bagging.py\", line 284, in fit\n    estimator.fit(X[:, features])\n  File \"C:\\src\\scripts\\mlu.py\", line 721, in outC\n    fb_model = FeatureBagging(n_estimators=1000,\n  File \"C:\\src\\scripts\\mlt.py\", line 468, in train\n    df, no = outC(raw_df, target, param.outlier_percentage)\n===\n   at Python.Runtime.PyObject.Invoke(PyTuple args, PyDict kw)\r\n   at Python.Runtime.PyObject.InvokeMethod(String name, PyTuple args, PyDict kw)\r\n   at Python.Runtime.PyObject.TryInvokeMember(InvokeMemberBinder binder, Object[] args, Object& result)\r\n   at CallSite.Target(Closure , CallSite , Object , String , Object )\r\n   at System.Dynamic.UpdateDelegates.UpdateAndExecute3[T0,T1,T2,TRet](CallSite site, T0 arg0, T1 arg1, T2 arg2)\r\n   at Intuition.ML.Trainer.Run(Guid tenantId, Trainparameter) in C:\\src\\src\\Intuition.AutoML\\Implementation\\Trainer.cs:line 91
```
If I increase the RAM it should work is it?, now RAM size is 32GB

**Comment by ameenmsit** [comment]
MemoryError: Unable to allocate 229. MiB for an array with shape (7936, 3785) and data type float64
I have only 8b RAM. What are my possible solutions.

**Comment by ameenmsit** [comment]
After uninstalling python 3.8 and install new anaconda with python 3.9 my problem resolved.

**Comment by Zanelehappybhengu** [comment]
> Is It solved? I am facing this problem in much lower level. i am just trying to create matrix cosine_sim = linear_kernel(tfmatrix, tfmatrix) and getting this - MemoryError: Unable to allocate 15.4 GiB for an array with shape (45466, 45466) and data type float64

same here i need help

**Comment by OrcunSamiTandogan** [comment]
We had to change something in there. I remember I have solved the issue after I did not need it anymore, but not remember how, now. However, changing, reinstalling the environment probably was the solution. Python version you use, must be composable with your operating system. And also, take care whether you are on kernel from Python or from Anaconda.

**Comment by Adusmichael** [comment]
MemoryError: Unable to allocate 6.87 GiB for an array with shape (36302, 25392) and data type float64

Am also getting this error. Am using Jupiter notebook. Please help

**Comment by Seqaeon** [comment]
MemoryError: Unable to allocate 2.40 TiB for an array with shape (7363219, 44840) and data type float64

**Comment by mirfan57** [comment]
MemoryError: Unable to allocate 3.31 GiB for an array with shape (44467, 10000) and data type float64

Getting this error while running cosine_similarity metric on vectors obtained from CountVectorizer.

**Comment by mans28in** [comment]
i am also getting this error :
Unable to allocate 6.54 GiB for an array with shape (146333, 6000) and data type float64

**Comment by CSIT-Knowledge-Hub-EBulk** [comment]
i am also getting this error :
MemoryError: Unable to allocate 10.7 GiB for an array with shape (71537, 20000) and data type float64

**Comment by Amit343** [comment]
Change the data types of your values in float

**Comment by Junaid432** [comment]
Me also face this error while ploting 9 second voice.
pls help me about this.
MemoryError: Unable to allocate 272. GiB for an array with shape (608580, 60002) and data type float64

**Comment by nbro10** [comment]
@ArchaeonSeq An array of shape (7363219, 44840), would have 330166739960 entries, each of which would require, let's say, 8 bytes (64 bits), so you would need 330166739960*8 bytes, which corresponds to ~2641333919 kilobytes, which corresponds to ~2641333 megabytes, which corresponds to 2641 GBs, which corresponds to 2.6 terabytes. No normal computer, as far as I know, has that amount of RAM. So, of course, you get that error.

**Comment by Rolilixs** [comment]
Hello!

I faced a similar problem, in my case I wanted to create a matrix of similarity between music artists based on their genres. And I got the following error:

```python
from sklearn.metrics.pairwise import cosine_similarity
similarity_matrix = cosine_similarity(<dataframe>)
```

`Unable to allocate 486. GiB for an array with shape (255473, 255473) and data type float64`

In my case, most of the values in the matrix will be zero. The solution to the problem was to use a sparse matrix from the **scipy** library.

```python
from scipy.sparse import csr_matrix
sparse_matrix = csr_matrix(<dataframe>)
```

After that, I was able to use the result to fill in the nearest neighbor model.
I'm not sure, but I think this will be a good solution only for matrices in which _most values are zero_.
